### ISSUE TITLE
Support Order and Range statistics

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#19729]
Order statistics as described in Wikipedia:
https://en.wikipedia.org/wiki/Order_statistic

We should create a class to support this in SymPy.

**Comment by Maelstrom6** [comment]
I tried doing it manually without creating any support for it and it was VERY compute intensive for some distributions. Any pdf's with a polynomial was fine but trying normal distributions became a nightmare. As an example, given the ordered normal iid sample {X(1), X(2), X(3), X(4), X(5)}, calculating `E(X(1) * X(3))` is not possible for sympy since it needed to integrate powers of `erf`. I think Wolfram could only go up to `erf(x)^3` before providing an unevaluated result.

I think iid random sampling functionality should be in place before we attempt order stats though.

**Comment by naveensaigit** [comment]
Currently, there is no way to declare i.i.d's in sympy. For this, I think the following approach might be good - 
i.i.d -> Order Statistics -> Range Statistics
First, to create an I.I.D, we can create a class `IID` in `sympy.stats` like this - 
``` python
>>> from sympy.stats import Normal, IID
>>> I = IID(Normal("N", 0, 1), 5)
>>> I[2]
I[2] // Independant Random Variable
>>> I(2)
I(2) // Order Statistic
```
I think we can use `RandomIndexedSymbol` ( just like in `DiscreteMarkovChain` and `ContinuousMarkovChain` ) using [ ] for each individual i.i.d and ( ) for the order statistic. A closed form pdf and cdf exists for continuous case. A closed form for discrete random variables with both finite and infinite support can be found [here](http://www.math.wm.edu/~leemis/2006informsjoc.pdf). 

Please let me know what you think of this approach. I can start working on it.
P.S - The paper I mentioned includes both cases with and without replacement. Do we need both of them in sympy?

**Comment by Upabjojr** [comment]
> I tried doing it manually without creating any support for it and it was VERY compute intensive for some distributions.

We can still use some tricks like caching or pre-computed results, if it can help.

> I think Wolfram could only go up to erf(x)^3 before providing an unevaluated result.

Unevaluated results are also fine. Sometimes you can still go on with unevaluated expressions.

**Comment by Upabjojr** [comment]
> Currently, there is no way to declare i.i.d's in sympy. For this, I think the following approach might be good -
> i.i.d -> Order Statistics -> Range Statistics

I think you can compute the order statistic of dependent random variables as well. Independence just makes the computation simpler.

**Comment by Maelstrom6** [comment]
> `I = IID(Normal("N", 0, 1), 5)`

I think that looks good. That's also what I was thinking.

The square brackets and round brackets are also very smart and should be used. I was also thinking of `OrderStat([X, Y, Z], 1)` where `X` `Y` and `Z` are user-defined independent rv's and not necessarily from an iid sample.

> The paper I mentioned includes both cases with and without replacement. Do we need both of them in sympy?

Rather leave it as with replacement. A sample without replacement is, by its very nature, dependent and so not iid.

---

> I think you can compute the order statistic of dependent random variables as well. Independence just makes the computation simpler.

It's probably best to just do it for iid rv's for now. It won't be too hard to extend it to independent random variables but I think it will be difficult to extend it to dependent random variables.

---

We will need a way of finding the joint distribution of these random vairbales without having to integrate. Conside:
```python
from sympy import *

X = IID(Normal('X', 0, 1), 1000)
P(X(0)+X(999) < 0)
```
Here, sympy should not do 998 integrals to find the answer. Rather we should use previous theorems:

Let {X1, ..., Xn} be an iid sample from a continuous distribution with pdf f(x).

Then the pdf of X(i) is given by:
`$g(x) = {n \choose i-1, 1, n-i} F(x)^{i-1} f(x) (1-F(x))^{n-i}$ `
![image](https://user-images.githubusercontent.com/45897212/96568483-4aa4b700-12c8-11eb-9e0c-e8fb0856b805.png)

where `$ {n \choose a, b, c}$` is the multinomial coefficient.
Notice how the powers match the entries in the multinomial coefficient and that the pdf's and cdf's kind of increase as we go along.

Also, the joint pdf of X(i) and X(j) is given by:
`$g(x, y) = {n \choose i-1, 1, j-i-1, 1, n-j}F(x)^{i-1} f(x) (F(y)-F(x))^{j-i-1} f(y) (1-F(y))^{n-j}$ `
![image](https://user-images.githubusercontent.com/45897212/96568603-6a3bdf80-12c8-11eb-89c1-e34fe89db664.png)

for x < y and both in the original support.

This can be extended to an arbitrary number of variables without having to ever do integration:
The joint pdf of X(i1), X(i2), ..., X(im) is given by:
`$g(x_1, x_2, ..., x_m) = {n \choose i_1-1, 1, i_2-i_1-1, 1, ..., i_m-i_{m-1}-1, 1, n-i_m} F(x_1)^{i_1-1} f(x_1) (F(x_2)-F(x_1))^{i_2-i_1-1} f(x_2) ... (F(x_{m-1}) - F(x_m))^{i_m-i_{m-1}-1} f(x_m) (1-F(x_m))^{n-i_m}$ `
![image](https://user-images.githubusercontent.com/45897212/96568684-82abfa00-12c8-11eb-9e82-7996a8fb1ea5.png)

for x_1 < x_2 < ..., < x_m and all in the original support.

I have no idea how to extend this for non-identical or dependent rv's.

**Comment by naveensaigit** [comment]
> I was also thinking of OrderStat([X, Y, Z], 1) where `X` `Y` and `Z` are user-defined independent rv's and not necessarily from an iid sample.

Sure, I haven't looked much into order statistics for dependent variables. It maybe more complex.

> Here, sympy should not do 998 integrals to find the answer. Rather we should use previous theorems

Yes, we should definitely add the joint distributions of order statistics. Is there a closed form for CDF of joint distributions as well? I was thinking we should use closed forms as much as possible because the time to compute CDF from the PDF everytime will be a lot or it might even hang since it would involve complex integrals.

> I have no idea how to extend this for non-identical or dependent rv's.

Yeah, we should go through some literature for this.

**Comment by naveensaigit** [comment]
> > Currently, there is no way to declare i.i.d's in sympy. For this, I think the following approach might be good -
> > i.i.d -> Order Statistics -> Range Statistics
> 
> I think you can compute the order statistic of dependent random variables as well. Independence just makes the computation simpler.

Yes, I was thinking we could add it for i.i.d's and later extend it to non-dependent ones.
